Simplify parsing untrusted code and strengthen DOMPurify config

rugk opened the pull request:
While the previous method was safe (a least I consider it safe), as only `args[0]` is being parsed as HTML, and we usually consider the translation files a safe source; I think this is way better to be fail-safe and test i for HTML in a way that does not require creating a "real" DOM element.

Disclosure: Coded with AI assistance.

---

> This method parses its input into a separate in-memory DOM, disabling any [<script>](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/script) elements and stopping event handlers from running. While the returned document is effectively inert, event handlers and scripts in its DOM will be able to run if they are inserted into the visible DOM.

https://developer.mozilla.org/en-US/docs/Web/API/DOMParser/parseFromString#security_considerations

Aka this is safer as only if you seem to use the resulting HTML element it does anything bad.

---

Additionally it re-uses the existing DOMPurify `purifyHtmlConfig` and only further strengthen it. This e.g. makes sure the `ALLOWED_URI_REGEXP` is respected aka only some safe links are allowed. 

--

Tested so that it basically all still seems to work:
<img width="1036" height="287" alt="image" src="https://github.com/user-attachments/assets/59f7e386-7244-440e-be99-14c71ea97dec" />

rugk committed:
refactor: use DOMParser for checking if translation is HTML

rugk committed:
refactor: use given HTML config for DOMPurify

rugk committed:
chore update SRI hash of main JS file

elrido commented on js/privatebin.js:
```diff
@@ -963,13 +962,17 @@ jQuery.PrivateBin = (function($) {
+                // If parsing fails for any reason, consider it not HTML to avoid
+                // treating arbitrary strings as markup.
+                return false;
+            }
```
Puh, this function keeps getting more and more complicated. Originally, before 0462da6, we simply checked for `args[0].indexOf('<a') !== -1` - no risk for any injection or execution of code and likely faster.

But we added `kbd` tags to the translations, so now there are message IDs with no anchors that still contain keyboard tags. How about we simplify this to:
```suggestion
            // message IDs are allowed to contain anchors, spans, keyboard and emphasis tags
            // we can recognize all of them by only checking for anchors and keyboard tags
            return args[0].indexOf('<a') !== -1 || args[0].indexOf('<kbd') !== -1;
```

elrido reviewed (changes requested):
The JS unit tests fail with an out of memory error, which indicates that all generated variants of a test fail (in some jsVerify based test). Likely the `DOMParser` API is not handled by the jsDOM mock in nodeJS. You'd need to find some shim to simulate it in the unit tests or disable the tests that trigger this.

I do like the idea of restricting the DOMpurify rules on these further (incl. the link checks). So hope the AI can help us get this into a state that passes our tests.

rugk commented on js/privatebin.js:
```diff
@@ -812,12 +812,11 @@ jQuery.PrivateBin = (function($) {
                 // only allow tags/attributes we actually use in translations
-                output = DOMPurify.sanitize(
-                    output, {
+                const sanitizeConfig = Object.assign({}, purifyHtmlConfig, {
```
a) yeah!

Good spot for the second one, I found:
> note that the USE_PROFILES setting will override the ALLOWED_TAGS setting

https://github.com/cure53/DOMPurify/?tab=readme-ov-file#control-our-allow-lists-and-block-lists

rugk committed:
refactor: drastically simplify JS "has HTML" extension again

Co-authored-by: El RIDO <elrido@gmx.net>

rugk committed:
refactor: use modern spread syntax for combining object

Given it's 2026 this really should be supported by all browsers now: https://caniuse.com/mdn-javascript_operators_spread,mdn-javascript_operators_spread_spread_in_arrays,mdn-javascript_operators_spread_spread_in_function_calls,mdn-javascript_operators_spread_spread_in_object_literals

rugk committed:
refactor: introduce purifyHtmlConfigStrictSubset

rugk committed:
Merge branch 'master' into xss/jsImprove

rugk committed:
wipfix: fix JS syntax errors

rugk reviewed (commented):
Okay now adjusted and tested and it still seems to work everything:

<img width="1165" height="429" alt="Image" src="https://github.com/user-attachments/assets/58b2495f-53e4-4990-8fbb-98f4d95e8d48" />

I took the liberty to also apply the stricter DOMPurify config to the link conversion, where it was also used.

rugk committed:
style: fix/remove trailing comma

rugk committed:
docs: improve JSDoc

rugk committed:
test: fix tests

elrido reviewed (approved):
Ah, nice, so the tests actually caught a regression. Thank you!

elrido commented:
Only detected in manual testing during the 2.0.4 release, that this strips the syntax highlighting markup. I've added an option parameter so we can disable the stricter settings just for that format in 14c8b2fa252602744b7eacbfa5175ff359149be5 - I would have preferred to get this done through PR and reviewed, but didn't want to delay the release by another week or weeks. :-(
